docs/src/prototypes/MenuButton/MenuButton.tsx

```diff
@@ -0,0 +1,198 @@
+import {
+  Accessibility,
+  AccessibilityBehavior,
+  Button,
+  ButtonProps,
+  Menu,
+  MenuProps,
+  Ref,
+  ShorthandValue,
+} from '@stardust-ui/react'
+import * as _ from 'lodash'
+import * as keyboardKey from 'keyboard-key'
+import * as PopperJS from 'popper.js'
+import * as React from 'react'
+import { Manager, Reference, Popper } from 'react-popper'
+
+import { focusButton, focusMenuItem, focusNearest } from './focusUtils'
+import menuButtonBehavior from './menuButtonBehavior'
+import MenuButtonWrapper from './MenuButtonWrapper'
+
+export interface MenuButtonProps {
+  accessibility?: Accessibility
+  button: ShorthandValue<ButtonProps>
+  buttonId: string
+  disabled?: boolean
+  menu: ShorthandValue<MenuProps>
+  menuId: string
+  placement?: PopperJS.Placement
+}
+
+export interface MenuButtonState {
+  lastKeyCode: null | number
+  lastShiftKey: boolean
+  menuOpen: boolean
+}
+
+class MenuButton extends React.Component<MenuButtonProps, MenuButtonState> {
+  static defaultProps = {
+    accessibility: menuButtonBehavior,
+    placement: 'bottom',
+  }
+
+  state: MenuButtonState = {
+    lastKeyCode: null,
+    lastShiftKey: false,
+    menuOpen: false,
+  }
+
+  buttonRef = React.createRef<HTMLButtonElement>()
+  menuRef = React.createRef<HTMLUListElement>()
+
+  componentDidUpdate(_, prevState: MenuButtonState) {
+    if (!prevState.menuOpen && this.state.menuOpen) {
+      document.addEventListener('click', this.handleDocumentClick)
+
+      focusMenuItem(this.menuRef, this.state.lastKeyCode === keyboardKey.ArrowUp ? 'last' : 'first')
+    }
+
+    if (prevState.menuOpen && !this.state.menuOpen) {
+      document.removeEventListener('click', this.handleDocumentClick)
+
+      switch (this.state.lastKeyCode) {
+        case keyboardKey.Enter:
+        case keyboardKey.Escape:
+          focusButton(this.buttonRef)
+          break
+
+        case keyboardKey.Tab:
+          focusNearest(this.buttonRef, this.state.lastShiftKey ? 'previous' : 'next')
+          break
+      }
+    }
+  }
+
+  handleButtonOverrides = (predefinedProps: ButtonProps) => ({
+    onClick: (e: React.SyntheticEvent, buttonProps: ButtonProps) => {
+      _.invoke(predefinedProps, 'onClick', e, buttonProps)
+      this.setState(prevState => ({
+        lastKeyCode: null,
+        menuOpen: !prevState.menuOpen,
+      }))
+    },
+  })
+
+  handleDocumentClick = (e: MouseEvent) => {
+    const { menuOpen } = this.state
+    const target = e.target as HTMLElement
+    const isInside =
+      _.invoke(this.buttonRef.current, 'contains', target) &&
```

Comment:
this should be `or` condition - actually, now it will never evaluate to `true`

Reply:
Great catch 👍